// @polkadot/ui-settings/cjs/Settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.settings = exports.Settings = void 0;
const tslib_1 = require("tslib");
const eventemitter3_1 = require("eventemitter3");
const store_1 = tslib_1.__importDefault(require("store"));
const util_1 = require("@polkadot/util");
const index_js_1 = require("./defaults/index.js");
function withDefault(options, option, fallback) {
    const _option = option || fallback;
    return options.some(({ value }) => value === _option)
        ? _option
        : fallback;
}
class Settings {
    #emitter;
    #apiType;
    // will become deprecated for supporting substrate connect light clients. apiType structure should be used instead
    #apiUrl;
    #camera;
    #i18nLang;
    #icon;
    #ledgerApp;
    #ledgerConn;
    #locking;
    #metadataUp;
    #prefix;
    #storage;
    #uiMode;
    #uiTheme;
    #notification;
    constructor() {
        const settings = store_1.default.get('settings') || {};
        this.#emitter = new eventemitter3_1.EventEmitter();
        // will become deprecated for supporting substrate connect light clients. apiType structure should be used instead
        this.#apiUrl = (typeof settings.apiUrl === 'string' && settings.apiUrl) || (util_1.hasProcess && process.env?.['WS_URL']) || index_js_1.ENDPOINT_DEFAULT.value;
        this.#apiType = { param: this.#apiUrl, type: 'json-rpc' };
        this.#camera = withDefault(index_js_1.CAMERA, settings.camera, index_js_1.CAMERA_DEFAULT);
        this.#ledgerApp = withDefault(index_js_1.LEDGER_APP, settings.ledgerApp, index_js_1.LEDGER_APP_DEFAULT);
        this.#ledgerConn = withDefault(index_js_1.LEDGER_CONN, settings.ledgerConn, index_js_1.LEDGER_CONN_DEFAULT);
        this.#i18nLang = settings.i18nLang || index_js_1.LANGUAGE_DEFAULT;
        this.#icon = settings.icon || index_js_1.ICON_DEFAULT;
        this.#locking = settings.locking || index_js_1.LOCKING_DEFAULT;
        this.#metadataUp = withDefault(index_js_1.METADATA_UP, settings.storage, index_js_1.METADATA_UP_DEFAULT);
        this.#notification = settings.notification || index_js_1.NOTIFICATION_DEFAULT;
        this.#prefix = (0, util_1.isUndefined)(settings.prefix) ? index_js_1.PREFIX_DEFAULT : settings.prefix;
        this.#storage = withDefault(index_js_1.STORAGE, settings.storage, index_js_1.STORAGE_DEFAULT);
        this.#uiMode = settings.uiMode || index_js_1.UIMODE_DEFAULT;
        this.#uiTheme = settings.uiTheme || index_js_1.UITHEME_DEFAULT;
    }
    get camera() {
        return this.#camera;
    }
    get apiType() {
        return this.#apiType;
    }
    get apiUrl() {
        return this.#apiUrl;
    }
    get i18nLang() {
        return this.#i18nLang;
    }
    get icon() {
        return this.#icon;
    }
    get notification() {
        return this.#notification;
    }
    get ledgerApp() {
        return this.#ledgerApp;
    }
    get ledgerConn() {
        return this.#ledgerConn;
    }
    get locking() {
        return this.#locking;
    }
    get metadataUp() {
        return this.#metadataUp;
    }
    get prefix() {
        return this.#prefix;
    }
    get storage() {
        return this.#storage;
    }
    get uiMode() {
        return this.#uiMode;
    }
    get uiTheme() {
        return this.#uiTheme;
    }
    get availableCamera() {
        return index_js_1.CAMERA;
    }
    get availableCryptos() {
        return index_js_1.CRYPTOS;
    }
    get availableCryptosEth() {
        return index_js_1.CRYPTOS_ETH;
    }
    get availableCryptosLedger() {
        return index_js_1.CRYPTOS_LEDGER;
    }
    get availableIcons() {
        return index_js_1.ICONS;
    }
    get availableLedgerApp() {
        return index_js_1.LEDGER_APP;
    }
    get availableLedgerConn() {
        return index_js_1.LEDGER_CONN;
    }
    get availableLocking() {
        return index_js_1.LOCKING;
    }
    get availableMetadataUp() {
        return index_js_1.METADATA_UP;
    }
    get availableNodes() {
        return index_js_1.ENDPOINTS;
    }
    get availablePrefixes() {
        return index_js_1.PREFIXES;
    }
    get availableStorage() {
        return index_js_1.STORAGE;
    }
    get availableUIModes() {
        return index_js_1.UIMODES;
    }
    get availableUIThemes() {
        return index_js_1.UITHEMES;
    }
    get() {
        return {
            apiType: this.#apiType,
            apiUrl: this.#apiUrl,
            camera: this.#camera,
            i18nLang: this.#i18nLang,
            icon: this.#icon,
            ledgerApp: this.#ledgerApp,
            ledgerConn: this.#ledgerConn,
            locking: this.#locking,
            metadataUp: this.#metadataUp,
            notification: this.#notification,
            prefix: this.#prefix,
            storage: this.#storage,
            uiMode: this.#uiMode,
            uiTheme: this.#uiTheme
        };
    }
    set(settings) {
        this.#apiType = settings.apiType || this.#apiType;
        this.#apiUrl = settings.apiUrl || this.#apiUrl;
        this.#camera = settings.camera || this.#camera;
        this.#ledgerConn = settings.ledgerConn || this.#ledgerConn;
        this.#ledgerApp = settings.ledgerApp || this.#ledgerApp;
        this.#i18nLang = settings.i18nLang || this.#i18nLang;
        this.#icon = settings.icon || this.#icon;
        this.#locking = settings.locking || this.#locking;
        this.#metadataUp = settings.metadataUp || this.#metadataUp;
        this.#notification = settings.notification || this.#notification;
        this.#prefix = (0, util_1.isUndefined)(settings.prefix) ? this.#prefix : settings.prefix;
        this.#storage = settings.storage || this.#storage;
        this.#uiMode = settings.uiMode || this.#uiMode;
        this.#uiTheme = settings.uiTheme || this.#uiTheme;
        const newValues = this.get();
        store_1.default.set('settings', newValues);
        this.#emitter.emit('change', newValues);
    }
    on(type, cb) {
        this.#emitter.on(type, cb);
    }
}
exports.Settings = Settings;
exports.settings = new Settings();
